### ISSUE TITLE
Merge penumbra-web-assembly into monorepo

### ISSUE THREAD

**Issue by redshiftzero** [penumbra-zone/penumbra#1990]
## Overview
We want to merge the external create https://github.com/zpoken/penumbra-web-assembly into this monorepo. 

## Next steps

- [x] Get upstream wasm-pack build logic from Zpoken, PR into this repo: https://github.com/penumbra-zone/penumbra/pull/2163
  - [x] Adapt our CI workflows to add `wasm32-unknown-unknown` target for the `penumbra-web-assembly` crate: https://github.com/penumbra-zone/penumbra/pull/2163/commits/7cc20770067265a66f5d1310918d060a89fe8b7d
  - [x] Identify problematic areas in our codebase, where our deps cause build to fail
  - [x] Plan refactors for async (particularly tokio) code, add feature-gating where possible to unblock wasm32 builds
  - [x] Implement code changes requested by Zpoken; taken from [related discussion in Discord](https://discord.com/channels/824484045370818580/979785365437165658/1075184618178875462): 
    - [x] remove  penumbra-chain -> penumbra-storage -> tonic dependency  - #1857
    - [x] remove  penumbra-chain -> penumbra-proto -> tonic dependency 
    - [x] make jmt optional for penumbra-chain and penumbra-crypto - https://github.com/penumbra-zone/penumbra/pull/2025
    - [x] remove  penumbra-transaction -> tokio dependency -  #1989
  - [x] Get all CI workflows passing on PR
- [x] Clean up Rust code:
  - [x] Delete the `wasm-compat` crate (assuming it's actually just for build testing?)
  - [x] Rename the `penumbra-web-assembly` crate to `penumbra-wasm` in a `/wasm` directory, to match the other crates (e.g., `penumbra-storage` is in `/storage`)
  - [x] Run `cargo fmt` on `penumbra-wasm`
  - [x] Add `cargo fmt` to the `penumbra-wasm` CI (assuming we keep separate CI)
- [ ] Merge PR into penumbra/main with CI guarding against WASM-breaking regressions

Later:
    - [ ] Optional: have the possibility to use in penumbra-web-assembly  struct from penumbra-view (SpendableNoteRecord, SwapRecord) 

## CI changes
We should have a CI check to prevent breaking downstream users that are compiling our crates to wasm

To do so:
1. enable the `wasm32-unknown-unknown` target using rustup 
2. build only the selected packages (at least penumbra-transaction, penumbra-crypto, maybe others?) with `--target wasm32-unknown-unknown`

Per comment [here](https://github.com/penumbra-zone/penumbra/pull/1989#issuecomment-1430030743) this may involving adding some features that could be added by downstream Rust code. 

Another option suggested by @hdevalence is to merge downstream Rust code (wasm-bindgen) into the main repo. Then CI would fail if we break the build of the downstream wasm-bindgen code.

**Comment by redshiftzero** [comment]
There are a bunch of build errors if we compile to `wasm32-unknown-unknown` (try out branch `wasm-builderrs`) so it may make sense to do the latter option - merging the wasm-bindgen stuff into the main repo - or add a periodic CI job that builds latest penumbra against `penumbra-web-assembly`

**Comment by redshiftzero** [comment]
From Discord next steps:
1. remove  penumbra-chain -> penumbra-storage -> tonic dependency 
2. remove  penumbra-chain -> penumbra-proto -> tonic dependency 
3. make jmt optional for penumbra-chain and penumbra-crypto 
4. DONE - remove  penumbra-transaction -> tokio dependency 
5. Optional: have the possibility to use in penumbra-web-assembly  struct from penumbra-view (SpendableNoteRecord, SwapRecord)

**Comment by conorsch** [comment]
We didn't assign this ticket during last sprint planning, but today at retro we decided it was worth driving forward. I'll take a look and see what we can get done quickly.

**Comment by conorsch** [comment]
Reformatted the OP based on today's discussion of goals. The good news is: several of the asked-for tasks were accomplished already, so I've cross-linked them and checked the boxes. There's one outstanding PR from today that seems unobjectionable (#2025). 

The bad news is, I could use some guidance on this point:

> remove penumbra-chain -> penumbra-proto -> tonic dependency

Is that ask feasible? As I understand it, `chain` _must_ depend on `proto`, and `proto` has a whole lot of tonic dependency. Do we want to try to gate the imports with features? 

I'll spend some time with [the downstream repo](https://github.com/zpoken/penumbra-web-assembly) and see if I can get more perspective on the changes required, or already implemented over there.

**Comment by hdevalence** [comment]
The `tonic` dependencies in `proto` are supposed to be gated on an `rpc` feature; disabling it _should_ remove everything with a `tonic` dependency.

**Comment by conorsch** [comment]
> build only the selected packages (at least penumbra-transaction, penumbra-crypto, maybe others?) 

On a spike, I was able to build the wasm target for the `crypto` and `proto` packages, but not the others, even when disabling default features. Many of the errors appeared to be related networking and socket-handling calls, which aren't available in wasm. Given how low our coverage is, I don't think it's worthwhile yet to add CI jobs to guard against breakage.

Addressing the initial ask in the OP, that of merging in the `penumbra-web-assembly` repo, I'm not convinced that's worthwhile at present, either: there's no mention of any build commands to generate the assets in that repo, such as `penumbra_web_assembly_bg.wasm`. I suggest that we work with Zpoken to understand the current build process for generating assets used in the web platforms, then we can document that process explicitly, as well as add related tests to our CI workflows.

**Comment by conorsch** [comment]
Discussed in retro today, and we'll ask for additional details on the build process backing the artifacts in the `penumbra-web-assembly` repo at the next standing meeting. Hopefully we can take that process and stick it into CI, once it's documented. 

> Given how low our coverage is, I don't think it's worthwhile yet to add CI jobs to guard against breakage.

While spiking on this, I created a branch, so I'm going to submit that as a PR once CI is happy with it. No rush to get it in the next testnet, since it's CI-only.

**Comment by conorsch** [comment]
We've discussed this with the Zpoken team, and they plan to PR their wasm build logic into this repository. That'll enable us to test directly against their build logic, rather than using a stand-in crate as in #2086. 

Marking as blocked for now, since we're waiting for that external PR. Will pick this back up once we have vetted logic to test against.

**Comment by conorsch** [comment]
Updated the OP with a checklist tracking next steps, as [recently discussed in Discord](https://discord.com/channels/824484045370818580/1044457038886998137/1087420479586959451). We now have a CI check on the PR branch that fails, giving us a target for full compatibility. Soon I'll follow up with a bit more detail on the specific dependencies that are proving problematic.

**Comment by conorsch** [comment]
Pruned some feature flags on the WIP PR, focusing on `tokio` and `ibc` dependencies.

<details>
<summary>
tokio deps before changes
</summary

```
$ cargo tree -i tokio
tokio v1.26.0
├── h2 v0.3.16
│   ├── hyper v0.14.25
│   │   ├── axum v0.6.11
│   │   │   └── tonic v0.8.3
│   │   │       └── ibc-proto v0.26.0
│   │   │           ├── ibc v0.29.0
│   │   │           │   ├── penumbra-chain v0.1.0 (/home/conor/src/penumbra/chain)
│   │   │           │   │   ├── penumbra-transaction v0.1.0 (/home/conor/src/penumbra/transaction)
│   │   │           │   │   │   └── penumbra-web-assembly v0.48.0 (/home/conor/src/penumbra/penumbra-web-assembly)
│   │   │           │   │   └── penumbra-web-assembly v0.48.0 (/home/conor/src/penumbra/penumbra-web-assembly)
│   │   │           │   ├── penumbra-proto v0.1.0 (/home/conor/src/penumbra/proto)
│   │   │           │   │   ├── penumbra-chain v0.1.0 (/home/conor/src/penumbra/chain) (*)
│   │   │           │   │   ├── penumbra-crypto v0.1.0 (/home/conor/src/penumbra/crypto)
│   │   │           │   │   │   ├── penumbra-chain v0.1.0 (/home/conor/src/penumbra/chain) (*)
│   │   │           │   │   │   ├── penumbra-proof-params v0.1.0 (/home/conor/src/penumbra/proof-params)
│   │   │           │   │   │   │   └── penumbra-transaction v0.1.0 (/home/conor/src/penumbra/transaction) (*)
│   │   │           │   │   │   ├── penumbra-transaction v0.1.0 (/home/conor/src/penumbra/transaction) (*)
│   │   │           │   │   │   └── penumbra-web-assembly v0.48.0 (/home/conor/src/penumbra/penumbra-web-assembly)
│   │   │           │   │   ├── penumbra-tct v0.1.0 (/home/conor/src/penumbra/tct)
│   │   │           │   │   │   ├── penumbra-chain v0.1.0 (/home/conor/src/penumbra/chain) (*)
│   │   │           │   │   │   ├── penumbra-crypto v0.1.0 (/home/conor/src/penumbra/crypto) (*)
│   │   │           │   │   │   ├── penumbra-transaction v0.1.0 (/home/conor/src/penumbra/transaction) (*)
│   │   │           │   │   │   └── penumbra-web-assembly v0.48.0 (/home/conor/src/penumbra/penumbra-web-assembly)
│   │   │           │   │   ├── penumbra-transaction v0.1.0 (/home/conor/src/penumbra/transaction) (*)
│   │   │           │   │   └── penumbra-web-assembly v0.48.0 (/home/conor/src/penumbra/penumbra-web-assembly)
│   │   │           │   └── penumbra-transaction v0.1.0 (/home/conor/src/penumbra/transaction) (*)
│   │   │           ├── penumbra-proto v0.1.0 (/home/conor/src/penumbra/proto) (*)
│   │   │           └── penumbra-transaction v0.1.0 (/home/conor/src/penumbra/transaction) (*)
│   │   ├── hyper-timeout v0.4.1
│   │   │   └── tonic v0.8.3 (*)
│   │   └── tonic v0.8.3 (*)
│   └── tonic v0.8.3 (*)
├── hyper v0.14.25 (*)
├── hyper-timeout v0.4.1 (*)
├── penumbra-storage v0.1.0 (/home/conor/src/penumbra/storage)
│   ├── penumbra-chain v0.1.0 (/home/conor/src/penumbra/chain) (*)
│   └── penumbra-proto v0.1.0 (/home/conor/src/penumbra/proto) (*)
├── tokio-io-timeout v1.2.0
│   └── hyper-timeout v0.4.1 (*)
├── tokio-stream v0.1.12
│   ├── penumbra-storage v0.1.0 (/home/conor/src/penumbra/storage) (*)
│   └── tonic v0.8.3 (*)
├── tokio-util v0.7.7
│   ├── h2 v0.3.16 (*)
│   ├── tonic v0.8.3 (*)
│   └── tower v0.4.13
│       ├── axum v0.6.11 (*)
│       └── tonic v0.8.3 (*)
├── tonic v0.8.3 (*)
└── tower v0.4.13 (*)
```
</details>

<details>
<summary>
tokio deps after changes
</summary>

```
tokio v1.26.0
├── penumbra-storage v0.1.0 (/home/conor/src/penumbra/storage)
│   ├── penumbra-chain v0.1.0 (/home/conor/src/penumbra/chain)
│   │   ├── penumbra-transaction v0.1.0 (/home/conor/src/penumbra/transaction)
│   │   │   └── penumbra-web-assembly v0.48.0 (/home/conor/src/penumbra/penumbra-web-assembly)
│   │   └── penumbra-web-assembly v0.48.0 (/home/conor/src/penumbra/penumbra-web-assembly)
│   ├── penumbra-proto v0.1.0 (/home/conor/src/penumbra/proto)
│   │   ├── penumbra-chain v0.1.0 (/home/conor/src/penumbra/chain) (*)
│   │   ├── penumbra-crypto v0.1.0 (/home/conor/src/penumbra/crypto)
│   │   │   ├── penumbra-chain v0.1.0 (/home/conor/src/penumbra/chain) (*)
│   │   │   ├── penumbra-proof-params v0.1.0 (/home/conor/src/penumbra/proof-params)
│   │   │   │   └── penumbra-transaction v0.1.0 (/home/conor/src/penumbra/transaction) (*)
│   │   │   ├── penumbra-transaction v0.1.0 (/home/conor/src/penumbra/transaction) (*)
│   │   │   └── penumbra-web-assembly v0.48.0 (/home/conor/src/penumbra/penumbra-web-assembly)
│   │   ├── penumbra-tct v0.1.0 (/home/conor/src/penumbra/tct)
│   │   │   ├── penumbra-chain v0.1.0 (/home/conor/src/penumbra/chain) (*)
│   │   │   ├── penumbra-crypto v0.1.0 (/home/conor/src/penumbra/crypto) (*)
│   │   │   ├── penumbra-transaction v0.1.0 (/home/conor/src/penumbra/transaction) (*)
│   │   │   └── penumbra-web-assembly v0.48.0 (/home/conor/src/penumbra/penumbra-web-assembly)
│   │   ├── penumbra-transaction v0.1.0 (/home/conor/src/penumbra/transaction) (*)
│   │   └── penumbra-web-assembly v0.48.0 (/home/conor/src/penumbra/penumbra-web-assembly)
│   └── penumbra-web-assembly v0.48.0 (/home/conor/src/penumbra/penumbra-web-assembly)
└── tokio-stream v0.1.12
    └── penumbra-storage v0.1.0 (/home/conor/src/penumbra/storage) (*)

```
</details>


If I'm interpreting that output correctly, looks like we should focus on the `storage` crate's use of tokio, and see if we can get the dependency backed down to optional.

**Comment by conorsch** [comment]
Ah, drat: had to back out some of those `ibc` feature tweaks due to failing smoke test. Will try again more selectively.

**Comment by hdevalence** [comment]
The storage crate definitely needs tokio, and that won't ever change. We should figure out why it's being depended on — logically, the wasm code in the client shouldn't need the storage implementation for the fullnode...

**Comment by hdevalence** [comment]
I was able to get a version of the PR working in https://github.com/penumbra-zone/penumbra/pull/2206 ; I duplicated the branch only because I couldn't figure out how to push to the remote.  I'll update the checklist above.

**Comment by conorsch** [comment]
>  Add `cargo fmt` to the penumbra-wasm CI (assuming we keep separate CI)

This is done. We did indeed duplicate the CI config for `cargo fmt`, which is required because we're using a separate workspace for the `penumbra-wasm` crate. It's not strictly necessary that we do so, but it seems a nice-to-have to confirm working builds of downstream crates. At any point in the future we can opt to pull the `penumbra-wasm` crate into the parent workspace.
